File: impl-go/probe.go
package main

// The `warrant-conformance/1` candidate contract (conformance/CONTRACT.md).
//
// `conformance` is this binary checking itself against files in this checkout.
// `probe` is the inverse: it makes this binary a CANDIDATE that an external
// runner — one that has never seen this repository — can drive. One JSON request
// on stdin, one JSON response on stdout, exit 0 whenever an answer was produced.
//
// The exit status is deliberately NOT the verdict. "I say this body is invalid"
// and "I crashed before deciding" are different facts, and a contract that
// encodes both as a nonzero exit lets a broken candidate look like a strict one.

import (
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
)

const probeProtocol = "1"

var (
	probeBaseClasses = []string{"capabilities", "canon", "validate", "blob-hash",
		"sig-message", "verify-sig", "parse"}
	probeSettlementClasses = []string{"verify-store", "ski-run"}
)

// probeUnsupported is returned for a class this implementation genuinely cannot
// answer. The runner scores it UNRUN — a distinct outcome from both pass and
// fail, because a skipped vector class proves nothing and must not look like one
// that passed.
type probeUnsupported struct{ why string }

func (e probeUnsupported) Error() string { return e.why }

func probeString(m map[string]any, key string) (string, error) {
	v, ok := m[key].(string)
	if !ok {
		return "", fmt.Errorf("input.%s must be a string", key)
	}
	return v, nil
}

func probeBytes(m map[string]any, key string) ([]byte, error) {
	s, err := probeString(m, key)
	if err != nil {
		return nil, err
	}
	return base64.StdEncoding.DecodeString(s)
}

func probeAnswer(req map[string]any) (map[string]any, error) {
	class, _ := req["class"].(string)
	input, _ := req["input"].(map[string]any)
	if input == nil {
		input = map[string]any{}
	}

	switch class {
	case "capabilities":
		classes := append(append([]string{}, probeBaseClasses...), probeSettlementClasses...)
		return map[string]any{
			"name":    "warrant-go (independent implementation)",
			"version": "body-format/0.2",
			"grade":   "settlement",
			"classes": classes,
		}, nil

	case "canon":
		body, ok := input["body"].(map[string]any)
		if !ok {
			return map[string]any{"error": "body is not a JSON object"}, nil
		}
		canon, err := canonicalJSON(body)
		if err != nil {
			return map[string]any{"error": err.Error()}, nil
		}
		id, err := warrantID(body)
		if err != nil {
			return map[string]any{"error": err.Error()}, nil
		}
		return map[string]any{
			"canon_hex":  hex.EncodeToString(canon),
			"warrant_id": id,
		}, nil

	case "validate":
		body, ok := input["body"].(map[string]any)
		if !ok {
			return map[string]any{"valid": false,
				"errors": []string{"body is not a JSON object"}}, nil
		}
		errs := validateBody(body)
		if errs == nil {
			errs = []string{}
		}
		return map[string]any{"valid": len(errs) == 0, "errors": errs}, nil

	case "blob-hash":
		data, err := probeBytes(input, "bytes_base64")
		if err != nil {
			return nil, err
		}
		return map[string]any{"hash": blobHash(data)}, nil

	case "sig-message":
		wid, err := probeString(input, "warrant_id")
		if err != nil {
			return nil, err
		}
		msg, ok := sigMessage(wid)
		if !ok {
			return map[string]any{"error": "WarrantID is not 64 lowercase hex characters"}, nil
		}
		return map[string]any{"message_hex": hex.EncodeToString(msg)}, nil

	case "verify-sig":
		wid, err := probeString(input, "warrant_id")
		if err != nil {
			return nil, err
		}
		key, err := probeString(input, "key")
		if err != nil {
			return nil, err
		}
		sg, err := probeString(input, "sig")
		if err != nil {
			return nil, err
		}
		sig := map[string]any{"actor": "probe", "key": key, "sig": sg}
		return map[string]any{"valid": verifySig(wid, sig)}, nil

	case "parse":
		data, err := probeBytes(input, "bytes_base64")
		if err != nil {
			return nil, err
		}
		if _, derr := decodeStrictJSON(data); derr != nil {
			return map[string]any{"ok": false, "error": derr.Error()}, nil
		}
		return map[string]any{"ok": true}, nil

	case "verify-store":
		dir, err := probeString(input, "store_path")
		if err != nil {
			return nil, err
		}
		if !isDir(filepath.Join(dir, "records")) {
			return map[string]any{"error": "not a store"}, nil
		}
		var errs, warns int
		if g, _ := input["grade"].(string); g == "settlement" {
			trust, _ := input["trust_config_path"].(string)
			var genesis []string
			if raw, ok := input["genesis"].([]any); ok {
				for _, g := range raw {
					if s, ok := g.(string); ok {
						genesis = append(genesis, s)
					}
				}
			}
			errs, warns = verifyDirSettlement(dir, trust, genesis, true, nil)
		} else {
			errs, warns = verifyDir(dir, true, nil, true)
		}
		return map[string]any{"errors": errs, "warnings": warns}, nil

	case "ski-run":
		check, err := probeBytes(input, "check_base64")
		if err != nil {
			return nil, err
		}
		blobs := map[string][]byte{}
		if raw, ok := input["blobs_base64"].(map[string]any); ok {
			// Keys are the runner's labels, not trusted addresses: the store is
			// keyed by what the bytes actually hash to, so a mislabelled blob
			// cannot smuggle itself in under another blob's address.
			names := make([]string, 0, len(raw))
			for k := range raw {
				names = append(names, k)
			}
			sort.Strings(names)
			for _, k := range names {
				s, ok := raw[k].(string)
				if !ok {
					return nil, fmt.Errorf("blobs_base64[%s] must be a string", k)
				}
				b, derr := base64.StdEncoding.DecodeString(s)
				if derr != nil {
					return nil, derr
				}
				blobs[blobHash(b)] = b
			}
		}
		checkHex := blobHash(check)
		blobs[checkHex] = check
		verdict, resultHash, spent, rerr := runSkiCheckFromStore(blobs, checkHex)
		if rerr != nil {
			return map[string]any{"error": rerr.Error()}, nil
		}
		return map[string]any{
			"verdict":          verdict,
			"result_node_hash": resultHash,
			"atp_spent":        spent,
		}, nil
	}
	return nil, probeUnsupported{fmt.Sprintf("class %q is not implemented by this candidate", class)}
}

// probeMain reads exactly one request and writes exactly one response.
func probeMain() int {
	raw, err := io.ReadAll(os.Stdin)
	if err != nil {
		fmt.Fprintln(os.Stderr, "probe:", err)
		return 2
	}
	req, err := decodeStrictJSON(raw)
	if err != nil {
		fmt.Fprintln(os.Stderr, "probe: malformed request:", err)
		return 2
	}
	if v, _ := req["warrant_conformance"].(string); v != probeProtocol {
		fmt.Fprintf(os.Stderr, "probe: unsupported request protocol %q\n", v)
		return 2
	}
	resp := map[string]any{"warrant_conformance": probeProtocol, "id": req["id"]}
	out, err := probeAnswer(req)
	if err != nil {
		if u, ok := err.(probeUnsupported); ok {
			resp["unsupported"] = u.why
		} else {
			fmt.Fprintln(os.Stderr, "probe:", err)
			return 2
		}
	} else {
		resp["output"] = out
	}
	enc := json.NewEncoder(os.Stdout)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(resp); err != nil {
		fmt.Fprintln(os.Stderr, "probe:", err)
		return 2
	}
	return 0
}

File: impl-rs/src/probe.rs
//! The `warrant-conformance/1` candidate contract (conformance/CONTRACT.md).
//!
//! `conformance` is this binary checking itself against files in this checkout.
//! `probe` is the inverse: it makes this binary a CANDIDATE an external runner
//! can drive without ever seeing this repository. One JSON request on stdin, one
//! JSON response on stdout, exit 0 whenever an answer was produced.
//!
//! This implementation is SPEC §6 BASE grade by design — no settlement, no key
//! state, no ski@v1 re-execution — so it declares `grade: "base"` and answers
//! `unsupported` for the settlement-grade classes. That asymmetry is real and
//! the contract exists partly to make it legible: a runner must report those
//! classes as UNRUN, never as passed and never as failed.

use super::*;
use std::io::Read;

pub const PROTOCOL: &str = "1";

const BASE_CLASSES: &[&str] = &[
    "capabilities",
    "canon",
    "validate",
    "blob-hash",
    "sig-message",
    "verify-sig",
    "parse",
    "verify-store",
];

fn b64_decode(s: &str) -> Option<Vec<u8>> {
    let mut out = Vec::new();
    let mut acc: u32 = 0;
    let mut bits = 0u32;
    let mut pad = 0usize;
    for ch in s.bytes() {
        if ch == b'\n' || ch == b'\r' {
            continue;
        }
        let v = match ch {
            b'A'..=b'Z' => ch - b'A',
            b'a'..=b'z' => ch - b'a' + 26,
            b'0'..=b'9' => ch - b'0' + 52,
            b'+' => 62,
            b'/' => 63,
            b'=' => {
                pad += 1;
                continue;
            }
            _ => return None,
        };
        if pad > 0 {
            return None; // data after padding
        }
        acc = (acc << 6) | v as u32;
        bits += 6;
        if bits >= 8 {
            bits -= 8;
            out.push((acc >> bits) as u8);
        }
    }
    Some(out)
}

fn s(v: &str) -> Json {
    Json::Str(v.to_string())
}

fn obj(pairs: Vec<(&str, Json)>) -> Json {
    let mut m = BTreeMap::new();
    for (k, v) in pairs {
        m.insert(k.to_string(), v);
    }
    Json::Object(m)
}

fn get<'a>(input: &'a Json, key: &str) -> Option<&'a Json> {
    as_obj(input).and_then(|m| m.get(key))
}

fn get_str<'a>(input: &'a Json, key: &str) -> Result<&'a str, String> {
    get(input, key)
        .and_then(as_str)
        .ok_or_else(|| format!("input.{key} must be a string"))
}

fn get_bytes(input: &Json, key: &str) -> Result<Vec<u8>, String> {
    b64_decode(get_str(input, key)?).ok_or_else(|| format!("input.{key} is not valid base64"))
}

/// Ok(Some(output)) = answered; Ok(None) = unsupported; Err = protocol failure.
fn answer(class: &str, input: &Json) -> Result<Option<Json>, String> {
    match class {
        "capabilities" => Ok(Some(obj(vec![
            ("name", s("warrant-rs (independent implementation)")),
            ("version", s("body-format/0.2")),
            ("grade", s("base")),
            (
                "classes",
                Json::Array(BASE_CLASSES.iter().map(|c| s(c)).collect()),
            ),
        ]))),

        "canon" => {
            let body = get(input, "body").ok_or("input.body is required")?;
            let mut bytes = Vec::new();
            match canon(body, &mut bytes) {
                Err(e) => Ok(Some(obj(vec![("error", Json::Str(e))]))),
                Ok(()) => Ok(Some(obj(vec![
                    ("canon_hex", Json::Str(encode_hex(&bytes))),
                    ("warrant_id", Json::Str(encode_hex(&sha256(&bytes)))),
                ]))),
            }
        }

        "validate" => {
            let body = get(input, "body").ok_or("input.body is required")?;
            let errs = if as_obj(body).is_some() {
                validate_body(body)
            } else {
                vec!["body is not a JSON object".to_string()]
            };
            Ok(Some(obj(vec![
                ("valid", Json::Bool(errs.is_empty())),
                ("errors", Json::Array(errs.into_iter().map(Json::Str).collect())),
            ])))
        }

        "blob-hash" => {
            let data = get_bytes(input, "bytes_base64")?;
            Ok(Some(obj(vec![(
                "hash",
                Json::Str(encode_hex(&sha256(&data))),
            )])))
        }

        "sig-message" => {
            let wid = get_str(input, "warrant_id")?;
            match sig_message(wid) {
                None => Ok(Some(obj(vec![(
                    "error",
                    s("WarrantID is not 64 lowercase hex characters"),
                )]))),
                Some(m) => Ok(Some(obj(vec![(
                    "message_hex",
                    Json::Str(encode_hex(&m)),
                )]))),
            }
        }

        "verify-sig" => {
            let wid = get_str(input, "warrant_id")?.to_string();
            let mut sig = BTreeMap::new();
            sig.insert("key".to_string(), s(get_str(input, "key")?));
            sig.insert("sig".to_string(), s(get_str(input, "sig")?));
            Ok(Some(obj(vec![(
                "valid",
                Json::Bool(verify_sig(&wid, &sig)),
            )])))
        }

        "parse" => {
            let data = get_bytes(input, "bytes_base64")?;
            // Strictly the parser's own answer. The "top level is an object"
            // rule is NOT applied here: the three reference implementations
            // enforce it at three different layers and agree on the outcome, so
            // the pack does not vector it and this probe must not invent a
            // stricter parser than the binary actually has.
            Ok(Some(match Parser::parse(&data) {
                Ok(_) => obj(vec![("ok", Json::Bool(true))]),
                Err(e) => obj(vec![("ok", Json::Bool(false)), ("error", Json::Str(e))]),
            }))
        }

        "verify-store" => {
            // SPEC §6, base grade. A settlement-grade request is REFUSED rather
            // than answered with base-grade counts: this binary derives no key
            // state and reads no trust configuration, so returning "0 errors"
            // here would report a verification that never happened — which is
            // the exact failure SPEC §12.3 forbids. Refusing makes the runner
            // score it UNRUN, and the grade line then says `base`, truthfully.
            if get(input, "grade").and_then(as_str) == Some("settlement") {
                return Ok(None);
            }
            let store = get_str(input, "store_path")?;
            match verify_counts(store, true) {
                None => Ok(Some(obj(vec![("error", s("not a store"))]))),
                Some((errs, warns)) => Ok(Some(obj(vec![
                    ("errors", Json::Int(errs.to_string())),
                    ("warnings", Json::Int(warns.to_string())),
                ]))),
            }
        }

        _ => Ok(None),
    }
}

pub fn main() -> ExitCode {
    let mut raw = Vec::new();
    if let Err(e) = std::io::stdin().read_to_end(&mut raw) {
        eprintln!("probe: {e}");
        return ExitCode::from(2);
    }
    let req = match Parser::parse(&raw) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("probe: malformed request: {e}");
            return ExitCode::from(2);
        }
    };
    if get(&req, "warrant_conformance").and_then(as_str) != Some(PROTOCOL) {
        eprintln!("probe: unsupported request protocol");
        return ExitCode::from(2);
    }
    // The contract requires `id` to be a string, so echoing it needs no clone
    // of an arbitrary value.
    let id = get(&req, "id").and_then(as_str).unwrap_or("").to_string();
    let class = get(&req, "class").and_then(as_str).unwrap_or("").to_string();
    let empty = Json::Object(BTreeMap::new());
    let input = get(&req, "input").unwrap_or(&empty);

    let mut resp = BTreeMap::new();
    resp.insert("warrant_conformance".to_string(), s(PROTOCOL));
    resp.insert("id".to_string(), Json::Str(id));
    match answer(&class, input) {
        Err(e) => {
            eprintln!("probe: {e}");
            return ExitCode::from(2);
        }
        Ok(None) => {
            resp.insert(
                "unsupported".to_string(),
                Json::Str(format!(
                    "class \"{class}\" is not implemented by this candidate \
                     (warrant-rs is SPEC §6 base grade)"
                )),
            );
        }
        Ok(Some(out)) => {
            resp.insert("output".to_string(), out);
        }
    }
    let mut bytes = Vec::new();
    if let Err(e) = canon(&Json::Object(resp), &mut bytes) {
        eprintln!("probe: {e}");
        return ExitCode::from(2);
    }
    println!("{}", String::from_utf8_lossy(&bytes));
    ExitCode::SUCCESS
}
